src/coreclr/src/vm/tailcallhelp.cpp

```diff
@@ -0,0 +1,871 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "corpriv.h"
+#include "tailcallhelp.h"
+#include "dllimport.h"
+#include "formattype.h"
+#include "sigformat.h"
+#include "gcrefmap.h"
+#include "threads.h"
+
+#ifndef CROSSGEN_COMPILE
+
+FCIMPL2(void*, TailCallHelp::AllocTailCallArgBuffer, INT32 size, void* gcDesc)
+{
+    FCALL_CONTRACT;
+
+    _ASSERTE(size >= 0);
+
```

Comment:
This can throw exceptions, so it needs `HELPER_METHOD_FRAME`

Reply:
Will the memory allocation exception automatically be translated to `OutOfMemoryException` on the managed side once this is in place?

Reply:
Yes

Reply:
This made release builds roughly 30% slower in microbenchmarks so I have changed it to use `nothrow` and to throw the managed exception manually, PTAL.

Reply:
Looks good.